### PR TITLE
Include epilogue guards in Dynamo code_parts

### LINKED ISSUE #137388: some guards are not shown in the code_parts
### 🐛 Describe the bug

```python
import torch

@torch.compile(backend="eager")
def f(x):
    if x.numel() >= 1024:
        x += 5
        return x
    else:
        x = x * 2
        return x

x = torch.ones(2,)
torch._dynamo.mark_dynamic(x, 0)
print(f(x)[0]) # tensor(2.)

from torch._dynamo.eval_frame import _debug_get_cache_entry_list, innermost_fn
cache_entries = _debug_get_cache_entry_list(innermost_fn(f))
cache_entry = cache_entries[0]
guard, code = cache_entry.check_fn, cache_entry.code
# the guard takes the local variables of an input frame, and tells whether a re-compilation should be triggered.
import dis
dis.dis(guard)
dis.dis(code)

for code_part in guard.code_parts:
    print(code_part)
```

It prints: 

```text
utils_device.CURRENT_DEVICE == None
___check_global_state()
___check_torch_function_mode_stack()
check_tensor(L['x'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=False, size=[None], stride=[1])
((L['x']._dynamo_dynamic_indices.issubset({0})) if hasattr(L['x'], '_dynamo_dynamic_indices') else True)
```

there's one missing. when I use `TORCH_TRACE` , the tree looks like:

```text
TREE_GUARD_MANAGER:
+- RootGuardManager
| +- DEFAULT_DEVICE: utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:460 in init_ambient_guards
| +- GLOBAL_STATE: ___check_global_state()
| +- GuardManager: source=L['x'], accessed_by=DictGetItemGuardAccessor(x)
| | +- TENSOR_MATCH: check_tensor(L['x'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=False, size=[None], stride=[1])
| | +- DYNAMIC_INDICES: ((L['x']._dynamo_dynamic_indices.issubset({0})) if hasattr(L['x'], '_dynamo_dynamic_indices') else True)
+- LAMBDA_GUARD: 2 <= L['x'].size()[0] <= 1023                                 # _dynamo/output_graph.py:452 in init_ambient_guards
```

we might only have `RootGuardManager` in `code_parts` , and the `LAMBDA_GUARD` part is missing.

related pr: https://github.com/pytorch/pytorch/pull/134181 (it adds the `code_parts` when cpp guard is used)

cc @ezyang @chauhang @penguinwu @bobrenjc93 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @rec @anijain2305 

### Versions

tested on 2.6.0.dev20241004

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.18.0) (oldest at bottom):
* __->__ #186222

GuardManagerWrapper.populate_code_parts_for_debugging() built its debug code_parts by walking normal leaf guards under the root manager and child managers. Shape guards installed as root epilogue lambda guards are stored separately, so the guard tree could show a LAMBDA_GUARD while guard_manager.code_parts omitted the same expression.

Collect leaf guard code parts through a shared helper and also visit root epilogue lambda guards. This keeps code_parts aligned with the guard tree without changing guard execution semantics. I chose to extend the existing debug population traversal rather than changing epilogue guard storage because the latter would affect guard ordering and runtime behavior.

Fixes #137388

Generated by my agent

Benchmark Results:

Command compared 5 repeats of 30 torch.compile calls for the issue's dynamic-shape repro on main and on this branch. Raw timings include first-run warmup.

main samples: [0.34694441698957235, 0.02792649803450331, 0.02704737603198737, 0.02789245598251, 0.027513677021488547]
main median: 0.02789245598251 seconds

fix samples: [0.972765775048174, 0.02747954602818936, 0.027007015014532954, 0.02609432302415371, 0.025623075023759156]
fix median: 0.027007015014532954 seconds

The median shows no compile-time regression for this repro-scale workload.

Test Plan:

python test/dynamo/test_guard_manager.py GuardManagerTests.test_code_parts_include_epilogue_lambda_guards

python - <<'PY' ... issue reproduction script ... PY

lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by anijain2305 — APPROVED** [review]

**Review by mlazos — APPROVED** [review]
